airflow/kubernetes/kube_client.py

```diff
@@ -105,16 +107,26 @@ def get_kube_client(
     if conf.getboolean("kubernetes_executor", "enable_tcp_keepalive"):
         _enable_tcp_keepalive()
 
+    new_client_config = Configuration.get_default_copy()
+    api_client_retry_configuration = conf.getjson("kubernetes", "api_client_retry_configuration", fallback={})
+
+    if not conf.getboolean("kubernetes_executor", "verify_ssl"):
+        _disable_verify_ssl()
+
+    if isinstance(api_client_retry_configuration, dict) and api_client_retry_configuration != {}:
+        new_client_config.retries = urllib3.util.Retry(**api_client_retry_configuration)
+    else:
+        raise ValueError("api_client_retry_configuration should be a dictionary")
```

Comment:
I think no need to raise an exception, we can just log an error message and skip configuring the client
```suggestion
        log.error("api_client_retry_configuration should be a dictionary.")
```

Reply:
Disagree, better fail fast than expect operators to read every single line of the log. What is the use case of allowing a partially invalid config file?

But if airflow already has some convention on how it deals with this in other places then that is better to follow.

Reply:
No such use case, but we can log an error and fallback to using the default settings instead. Will that be a better idea @hussein-awala @hterik?

Reply:
Why would you want to do that? If the config is structurally invalid it is better to notify the admin sooner rather than allow the misconfiguration to slip through and only be noticed much later, if it's noticed at all.

Reply:
Makes sense. What do you suggest @hussein-awala @hterik. I would think that having an exception was the best way but the unit tests dont align. Any thoughts? I am not able to figure out how to get the UTs working too with the exception code

Reply:
Do you have example of how the unit test is failing? I can't figure out how to find the github actions for prior commits of a PR.

Reply:
@hterik If I make the code to have an exception (state before the last commit), the tests look like this
```
/Users/adesai/.pyenv/versions/airflow-env/bin/python "/Users/adesai/Library/Application Support/JetBrains/GoLand2021.3/plugins/python-ce/helpers/pycharm/_jb_pytest_runner.py" --target test_client.py::TestClient
Testing started at 8:45 PM ...
Launching pytest with arguments test_client.py::TestClient --no-header --no-summary -q in /Users/adesai/Documents/OpenSource/Airflow/airflow/tests/kubernetes

============================= test session starts ==============================
collecting ... collected 7 items

test_client.py::TestClient::test_load_cluster_config ========================= AIRFLOW ==========================
Home of the user: /Users/adesai
Airflow home /Users/adesai/airflow
Skipping initializing of the DB as it was initialized already.
You can re-initialize the database by adding --with-db-init flag when running tests.
FAILED              [ 14%]
tests/kubernetes/test_client.py:30 (TestClient.test_load_cluster_config)
self = <tests.kubernetes.test_client.TestClient object at 0x112250df0>
config = <MagicMock name='config' id='4602299968'>

    @mock.patch("airflow.kubernetes.kube_client.config")
    def test_load_cluster_config(self, config):
>       get_kube_client(in_cluster=True)

test_client.py:33: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

in_cluster = True, cluster_context = None, config_file = None

    def get_kube_client(
        in_cluster: bool = conf.getboolean("kubernetes_executor", "in_cluster"),
        cluster_context: str | None = None,
        config_file: str | None = None,
    ) -> client.CoreV1Api:
        """
        Retrieves Kubernetes client.
    
        :param in_cluster: whether we are in cluster
        :param cluster_context: context of the cluster
        :param config_file: configuration file
        :return kubernetes client
        :rtype client.CoreV1Api
        """
        if not has_kubernetes:
            raise _import_err
    
        if conf.getboolean("kubernetes_executor", "enable_tcp_keepalive"):
            _enable_tcp_keepalive()
    
        new_client_config = Configuration.get_default_copy()
        api_client_retry_configuration = conf.getjson("kubernetes", "api_client_retry_configuration", fallback={})
    
        if not conf.getboolean("kubernetes_executor", "verify_ssl"):
            _disable_verify_ssl()
    
        if isinstance(api_client_retry_configuration, dict) and api_client_retry_configuration != {}:
            new_client_config.retries = urllib3.util.Retry(**api_client_retry_configuration)
        else:
>           raise ValueError("api_client_retry_configuration should be a dictionary")
E           ValueError: api_client_retry_configuration should be a dictionary

../../airflow/kubernetes/kube_client.py:119: ValueError
FAILED                 [ 28%]
tests/kubernetes/test_client.py:36 (TestClient.test_load_file_config)
self = <tests.kubernetes.test_client.TestClient object at 0x112250d00>
config = <MagicMock name='config' id='4603967328'>

    @mock.patch("airflow.kubernetes.kube_client.config")
    def test_load_file_config(self, config):
>       get_kube_client(in_cluster=False)

test_client.py:39: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

in_cluster = False, cluster_context = None, config_file = None

    def get_kube_client(
        in_cluster: bool = conf.getboolean("kubernetes_executor", "in_cluster"),
        cluster_context: str | None = None,
        config_file: str | None = None,
    ) -> client.CoreV1Api:
        """
        Retrieves Kubernetes client.
    
        :param in_cluster: whether we are in cluster
        :param cluster_context: context of the cluster
        :param config_file: configuration file
        :return kubernetes client
        :rtype client.CoreV1Api
        """
        if not has_kubernetes:
            raise _import_err
    
        if conf.getboolean("kubernetes_executor", "enable_tcp_keepalive"):
            _enable_tcp_keepalive()
    
        new_client_config = Configuration.get_default_copy()
        api_client_retry_configuration = conf.getjson("kubernetes", "api_client_retry_configuration", fallback={})
    
        if not conf.getboolean("kubernetes_executor", "verify_ssl"):
            _disable_verify_ssl()
    
        if isinstance(api_client_retry_configuration, dict) and api_client_retry_configuration != {}:
            new_client_config.retries = urllib3.util.Retry(**api_client_retry_configuration)
        else:
>           raise ValueError("api_client_retry_configuration should be a dictionary")
E           ValueError: api_client_retry_configuration should be a dictionary

../../airflow/kubernetes/kube_client.py:119: ValueError
FAILED          [ 42%]
tests/kubernetes/test_client.py:42 (TestClient.test_load_config_disable_ssl)
self = <tests.kubernetes.test_client.TestClient object at 0x112250e20>
conf = <MagicMock name='conf' id='4604005248'>
config = <MagicMock name='config' id='4604090400'>

    @mock.patch("airflow.kubernetes.kube_client.config")
    @mock.patch("airflow.kubernetes.kube_client.conf")
    def test_load_config_disable_ssl(self, conf, config):
        conf.getboolean.return_value = False
>       get_kube_client(in_cluster=False)

test_client.py:47: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

in_cluster = False, cluster_context = None, config_file = None

    def get_kube_client(
        in_cluster: bool = conf.getboolean("kubernetes_executor", "in_cluster"),
        cluster_context: str | None = None,
        config_file: str | None = None,
    ) -> client.CoreV1Api:
        """
        Retrieves Kubernetes client.
    
        :param in_cluster: whether we are in cluster
        :param cluster_context: context of the cluster
        :param config_file: configuration file
        :return kubernetes client
        :rtype client.CoreV1Api
        """
        if not has_kubernetes:
            raise _import_err
    
        if conf.getboolean("kubernetes_executor", "enable_tcp_keepalive"):
            _enable_tcp_keepalive()
    
        new_client_config = Configuration.get_default_copy()
        api_client_retry_configuration = conf.getjson("kubernetes", "api_client_retry_configuration", fallback={})
    
        if not conf.getboolean("kubernetes_executor", "verify_ssl"):
            _disable_verify_ssl()
    
        if isinstance(api_client_retry_configuration, dict) and api_client_retry_configuration != {}:
            new_client_config.retries = urllib3.util.Retry(**api_client_retry_configuration)
        else:
>           raise ValueError("api_client_retry_configuration should be a dictionary")
E           ValueError: api_client_retry_configuration should be a dictionary

../../airflow/kubernetes/kube_client.py:119: ValueError
FAILED             [ 57%]
tests/kubernetes/test_client.py:55 (TestClient.test_enable_tcp_keepalive)
self = <tests.kubernetes.test_client.TestClient object at 0x112250100>

    def test_enable_tcp_keepalive(self):
        socket_options = [
            (socket.SOL_SOCKET, socket.SO_KEEPALIVE, 1),
>           (socket.IPPROTO_TCP, socket.TCP_KEEPIDLE, 120),
            (socket.IPPROTO_TCP, socket.TCP_KEEPINTVL, 30),
            (socket.IPPROTO_TCP, socket.TCP_KEEPCNT, 6),
        ]
E       AttributeError: module 'socket' has no attribute 'TCP_KEEPIDLE'

test_client.py:59: AttributeError
PASSED               [ 71%]PASSED [ 85%][2023-03-14T20:45:44.378+0530] {configuration.py:1372} INFO - Reading default test configuration from /Users/adesai/Documents/OpenSource/Airflow/airflow/airflow/config_templates/default_test.cfg
[2023-03-14T20:45:44.380+0530] {configuration.py:1375} INFO - Reading test configuration from /Users/adesai/airflow/unittests.cfg
PASSED [100%]















test_client.py::TestClient::test_load_file_config 
test_client.py::TestClient::test_load_config_disable_ssl 
test_client.py::TestClient::test_enable_tcp_keepalive 
test_client.py::TestClient::test_disable_verify_ssl 
test_client.py::TestClient::test_api_client_retry_configuration_default 
test_client.py::TestClient::test_api_client_retry_configuration_correct_values 

========================= 4 failed, 3 passed in 0.89s ==========================

Process finished with exit code 1
```

Reply:
Put a breakpoint where it is raising the exception or add the actual type+value as part of the exception message, to help with the debugging and see what you actually get back.

I suspect it's the `and api_client_retry_configuration != {}` that should be exempt from the exception-condition. Because that is the value you get as fallback when the config is not provided. Missing value is still a valid configuration and should be allowed with fallback, not the same as value-present-but-invalid :)

Reply:
Hi @hterik the problem is that, for unit tests such as `test_load_config_disable_ssl`, a mock `conf` is used which is of type: `<class 'unittest.mock.MagicMock'>` resulting in the condition `isinstance(api_client_retry_configuration, dict)` being false. Maybe I should have a check for unit test so that this portion is ok. Let me try a simple solution here so that mock class as well as regular valid value is accepted